### PR TITLE
#590 - Renamed chooser to visitor and added function for parsed macro…

### PR DESCRIPTION
**Pull request by ambaxter**
How would you suggest testing this?

### CODE REVIEW

**Comment by emilio** [comment]
You can test it in the `bindgen-integration` crate. Should be straight-forward. Just create a visitor and add a macro in `cpp/Test.h` :)

File: src/lib.rs [hunk 1]
```diff
@@ -447,8 +447,8 @@ impl Builder {
 
     /// Allows configuring types in different situations, see the `TypeChooser`
     /// documentation.
-    pub fn type_chooser(mut self, cb: Box<chooser::TypeChooser>) -> Self {
-        self.options.type_chooser = Some(cb);
+    pub fn type_chooser(mut self, cb: Box<visitor::BindgenVisitor>) -> Self {
```
Comment (fitzgen) [inline src/lib.rs]: Why is this not renamed too?

**Comment by fitzgen** [comment]
Apologies for bike shedding, but I'm not in love with the name `BindgenVisitor`, and although I don't have any great replacements, we do already have visitors internally for dealing with libclang, so "visitor" is now an ambiguous term.

Spitting out anything that comes to mind:

* `TypeVisitor` is at least a little more specific
* `OnParse`
* `OnType`
* `ParseCallbacks`
* `ParseGuide`
* `TypeBehavior`
* `TypeCustomizer`
* `ParseBehavior`
* `ParseCustomizer`

Any other dieas?

File: src/lib.rs (hunk 1 above)
Comment (ambaxter) [inline src/lib.rs]: Fixed

**Comment by ambaxter** [comment]
I'm liking onParse or ParseCallbacks, myself.

**Comment by emilio** [comment]
ParseCallbacks SGTM

File: bindgen-integration/build.rs [hunk 2]
```diff
@@ -1,16 +1,33 @@
...
+use std::collections::HashSet;
 use std::env;
 use std::path::PathBuf;
+use std::sync::{Arc, RwLock};
 use bindgen::Builder;
+use bindgen::visitor::BindgenVisitor;
+
+#[derive(Debug)]
+struct MacroVisitor {
+ macros: Arc<RwLock<HashSet<String>>>,
```
Comment (emilio) [inline bindgen-integration/build.rs]: nit: indentation is off.

File: bindgen-integration/build.rs [hunk 3]
```diff
@@ -1,16 +1,33 @@
...
+}
 
 fn main() {
     gcc::Config::new()
         .cpp(true)
         .file("cpp/Test.cc")
         .compile("libtest.a");
 
+    let macros = Arc::new(RwLock::new(HashSet::new()));
+
```
Comment (emilio) [inline bindgen-integration/build.rs]: nit: extra newline.

Comment (ambaxter) [inline bindgen-integration/build.rs]: Fixed

File: bindgen-integration/build.rs (hunk 2 above)
Comment (ambaxter) [inline bindgen-integration/build.rs]: Fixed

**Comment by emilio** [comment]
This looks fine to me renamed to `ParseCallbacks` and squashed. Thanks!

**Comment by ambaxter** [comment]
Squashed!

**Comment by emilio** [comment]
@bors-servo r+

Thanks!

**Pull request merged**